config/configtelemetry/doc.go

```diff
@@ -4,4 +4,46 @@
 // Package configtelemetry defines various telemetry level for configuration.
 // It enables every component to have access to telemetry level
 // to enable metrics only when necessary.
+//
+// 1. configtelemetry.None
+//
+// No telemetry data should be collected.
+//
+// 2. configtelemetry.Basic
+//
+// Signals associated with this level cover the essential coverage of the component telemetry.
+//
+// This is the default level recommended when running the collector.
+//
+// The signals associated with this level must show low cardinality, the number of combinations of dimension values. The maximum cardinality allowed is set to 50.
```

Comment:
How are we going to enforce all the numbers in this doc?

Reply:
The issue says to offer guidance, not enforcement.

Before we go there, is this guidance any good? WDYT?

Reply:
We don't need to implement enforcement on this PR or have a detailed plan of how to do it, but we need to know that these are reasonably easy to enforce in practice, otherwise the guidance is not actionable

Reply:
It looks reasonably easy to enforce in practice to me. The only tricky thing I guess is to calculate the cardinality of the metric, by computing the total set of combinations that can be formed based on the dimensions of the metric. Is that what needs to be detailed in there?

Reply:
We should be able to measure the numbers easily, yes, my concern is more what are we going to do if we start deviating from these numbers. Do we get to a point where we just can't add more metrics at a certain level (since removing metrics is presumably a breaking change)? If a library that offers built-in instrumentation and starts producing many more metrics, what are we going to do?

To some extent this is a problem that we are going to have to solve regardless of what we put in here, but I think putting stringent requirements like these without having a good idea of how we are going we make our lives unnecessarily harder. I am tempted to just remove the numbers or make the numbers a rough guidance/approximation instead of a must-follow rule.

Reply:
> We should be able to measure the numbers easily, yes, my concern is more what are we going to do if we start deviating from these numbers. Do we get to a point where we just can't add more metrics at a certain level (since removing metrics is presumably a breaking change)? If a library that offers built-in instrumentation and starts producing many more metrics, what are we going to do?

We will need to reject the library in any case.

If we use say the Go SDK, and all of a sudden, it emits by default all sorts of metrics related to Go GC, this is in itself a change that can be considered a breaking change because it may break folks backends with high cardinality metrics.
> 
> To some extent this is a problem that we are going to have to solve regardless of what we put in here, but I think putting stringent requirements like these without having a good idea of how we are going we make our lives unnecessarily harder. I am tempted to just remove the numbers or make the numbers a rough guidance/approximation instead of a must-follow rule.

Yes, this whole doc is a guidance, I didn't mean for those numbers to feel such a hard rule. The numbers are what Pavol wanted to see, I am resistant also to provide absolute guidelines but offering numbers seemed to help enshrine some of the thinking.

Reply:
> Yes, this whole doc is a guidance, I didn't mean for those numbers to feel such a hard rule

Can we made this more explicit in the text? (e.g. saying the limits are 'best effort' or adding a 'roughly' here or there to make it clear that this is a guideline but not a hard rule).

> offering numbers seemed to help enshrine some of the thinking.

Yup, I agree it makes the distinction between the levels clearer

Reply:
Could we offer the numbers only as an example? e.g. "A typical component might have the following signals at this level."

IMO we're better off focusing the language on helping component developers make decisions about whether a metric or span should be included at a certain level, similar to how we might guide developers how to choose an appropriate log severity. (Also see my other comment about logs not fitting into this guidance.)

Reply:
I have revised the text to take into account your considerations, please review.